MG - 1529 - User Removal

rodneyosodo opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `MG-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

This is a feature because it adds user removal

<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?

This PR adds a new method to delete a user with a given ID in the repository package. The implementation of this method uses an SQL query to delete the client from the database.

Additionally, an extra method has been added to the gRPC auth service to handle entity deletion. This is useful because it reduces network trips back and forth from the client microservices to the auth microservice. Test cases have been added for the delete method to ensure its functionality. 


<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

- Resolves #1529 

## Have you included tests for your changes?

Yes, I have included tests for my changes.

<!--If you have not included tests, please explain why.
For example:
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?

Yes, I have updated the documentation for the new feature.

<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->

rodneyosodo committed:
feat(repository): Add delete method for clients

This commit adds a new method to delete a client with a given ID in the repository package. The implementation of this method uses a SQL query to delete the client from the database.

Additionally, test cases have been added for the delete method to ensure its functionality. These test cases include assertions for handling expected errors.

These changes enhance the functionality of the codebase by allowing clients to be deleted from the database.

Note: This commit does not include any breaking changes or additional modifications to other files.
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat(cli): add delete command

This commit adds a new command to the CLI for deleting a user. The command can be accessed by running `cli delete <user_id>`, where `<user_id>` is the ID of the user to be deleted.

The code changes include the addition of a new method in the SDK interface, `deleteUser`, which constructs a URL and makes a HTTP DELETE request to delete a user. Additionally, a mock implementation of the method has been included for testing purposes.

These changes enhance the functionality of the application by allowing users to delete user records directly from the CLI.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat(users): add delete endpoint

- Add new handlers for deleting and changing the status of a client in the users API
- Improve logging in the users API
- Update response types in the users API
- Update the Service interface in the users API

Add new functionality for deleting a client:
- Add a constant, event type, and method in the eventStore
- Implement the DeleteClient method in the service file
- The DeleteClient method disables a client by deleting its policies and checking for authorization
- Trace the operation by starting a span and calling the DeleteClient function of the wrapped service.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat(users): add delete endpoint

- Add new handlers for deleting and changing the status of a client in the users API
- Improve logging in the users API
- Update response types in the users API
- Update the Service interface in the users API

Add new functionality for deleting a client:
- Add a constant, event type, and method in the eventStore
- Implement the DeleteClient method in the service file
- The DeleteClient method disables a client by deleting its policies and checking for authorization
- Trace the operation by starting a span and calling the DeleteClient function of the wrapped service.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat(auth): add delete entity's policies

Add a method to delete entity's policies for example things or users. This helps reduced network communication between users and auth service

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
refactor(auth): add group policy deletion to auth

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

arvindh123 commented on users/service.go:
```diff
@@ -439,6 +439,33 @@ func (svc service) changeClientStatus(ctx context.Context, token string, client
+	}
+	if err := svc.checkSuperAdmin(ctx, res.GetId()); err != nil {
+		return err
+	}
```
@rodneyosodo We are planning like, only super admins can delete users

rodneyosodo replied:
Yes. It is only super admins who can delete users

rodneyosodo replied:
Should I also add that a user can be able to delete themselves?

arvindh123 replied:
We can add that later when required

arvindh123 commented on auth/service.go:
```diff
@@ -956,3 +961,122 @@ func DecodeDomainUserID(domainUserID string) (string, string) {
+			return err
+		}
+
+		return nil
```
For `svc.DeletePolicy` only object Type is required field. 
So the below snippet will delete all type of subjects related to thing object
```suggestion
		// Remove all policies related to thing
		req := PolicyReq{
			Object:     id,
			ObjectType: ThingType,
		}

		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}
		return nil
```

rodneyosodo replied:
This doesn't work. I get 
```json
{"time":"2024-03-21T08:36:41.624438869Z","level":"WARN","msg":"Delete entity policies failed to complete successfully","duration":"284.946µs","entity_type":"thing","id":"feae8203-452b-4f7c-a003-76b475e43b56","error":{"error":"malformed entity specification","message":"failed to remove the policies"}}
```

arvindh123 commented on auth/service.go:
```diff
@@ -956,3 +961,122 @@ func DecodeDomainUserID(domainUserID string) (string, string) {
+			}
+			if err := svc.agent.DeletePolicy(ctx, policy); err != nil {
+				return err
+			}
```
This below snippet will delete all Thing, Group, Channel relation related to user id within the domain
```suggestion
			policy := PolicyReq{
				Subject:     EncodeDomainUserID(domain.ID, id),
				SubjectType: UserType,
				ObjectType:  ThingType,
			}
			if err := svc.agent.DeletePolicy(ctx, policy); err != nil {
				return err
			}
			policy = PolicyReq{
				Subject:     EncodeDomainUserID(domain.ID, id),
				SubjectType: UserType,
				ObjectType:  GroupType,
			}
			if err := svc.agent.DeletePolicy(ctx, policy); err != nil {
				return err
			}

			policy = PolicyReq{
				Subject:     EncodeDomainUserID(domain.ID, id),
				SubjectType: UserType,
				Object:      domain.ID,
				ObjectType:  DomainType,
			}
			if err := svc.agent.DeletePolicy(ctx, policy); err != nil {
				return err
			}
```

arvindh123 commented on auth/service.go:
```diff
@@ -956,3 +961,122 @@ func DecodeDomainUserID(domainUserID string) (string, string) {
+		req.ObjectType = DomainType
+		if err := svc.agent.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
```
There is no need to remove policy between user_id and things/groups, because,There will be no subject just with user id in spicedb. 
User's Subject id in spicesdb will be always combination <domain_id>_<user_id>

```suggestion
		req := PolicyReq{
			Subject:     id,
			SubjectType: UserType,
			ObjectType:  PlatformType,
		    req.Object = MagistralaObject
		}
		if err := svc.agent.DeletePolicy(ctx, req); err != nil {
			return err
		}
```

rodneyosodo replied:
If we leave it spiceDB will see it as a valid policy and will not clean up

arvindh123 reviewed (changes requested):
Hello @rodneyosodo left some comments for delete policies in auth service.

rodneyosodo committed:
fix(cli): add host and invitations url to config

Check if domainsURL, invitationsURL and HostURL are empty to default to defURL

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
fix(auth): remove domain policies for groups and things

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
add users api docs for delete user

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
add tests

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
fix after rebase

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat(users): add delete process as background job

Add users as a background job after a specified period of being deleted

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

arvindh123 commented on auth/service.go:
```diff
@@ -1030,3 +1035,132 @@ func DecodeDomainUserID(domainUserID string) (string, string) {
+			return err
+		}
+
+		return nil
```
```suggestion
	case ThingType:
		// Remove policy of groups
		req := PolicyReq{
			Object:     id,
			ObjectType: ThingType,
		}

		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}
		return nil
```

The above suggestin will work , if we have changes in svc.agent.DeletePolicy show below 

```
func (pa *policyAgent) DeletePolicy(ctx context.Context, pr auth.PolicyReq) error {
	req := &v1.DeleteRelationshipsRequest{
		RelationshipFilter: &v1.RelationshipFilter{
			ResourceType:       pr.ObjectType,
			OptionalResourceId: pr.Object,
		},
	}

	if pr.Relation != "" {
		req.RelationshipFilter.OptionalRelation = pr.Relation
	}

	if pr.SubjectType != "" {
		req.RelationshipFilter.OptionalSubjectFilter = &v1.SubjectFilter{
			SubjectType: pr.SubjectType,
		}
		if pr.Subject != "" {
			req.RelationshipFilter.OptionalSubjectFilter.OptionalSubjectId = pr.Subject
		}
		if pr.SubjectRelation != "" {
			req.RelationshipFilter.OptionalSubjectFilter.OptionalRelation = &v1.SubjectFilter_RelationFilter{
				Relation: pr.SubjectRelation,
			}
		}
	}

	if _, err := pa.permissionClient.DeleteRelationships(ctx, req); err != nil {
		return errors.Wrap(errRemovePolicies, handleSpicedbError(err))
	}
	return nil
}
```

arvindh123 commented on auth/service.go:
```diff
@@ -1030,3 +1035,132 @@ func DecodeDomainUserID(domainUserID string) (string, string) {
+		req.SubjectType = UserType
+		if err := svc.DeletePolicy(ctx, req); err != nil {
+			return err
+		}
```
```suggestion
	case GroupType:
		// Remove policies where group is subject and object is another child group.
		req := PolicyReq{
			SubjectType: GroupType,
			Subject:     id,
			ObjectType:  GroupType,
		}
		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}

		// Remove policies where group is subject and object is thing.
		req.ObjectType = ThingType
		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}

		// Remove all policies where group is object.
		req = PolicyReq{
			Object:     id,
			ObjectType: GroupType,
		}
		if err := svc.DeletePolicy(ctx, req); err != nil {
			return err
		}
```

The above suggestin will work , if we have changes in svc.agent.DeletePolicy show below 

```
func (pa *policyAgent) DeletePolicy(ctx context.Context, pr auth.PolicyReq) error {
	req := &v1.DeleteRelationshipsRequest{
		RelationshipFilter: &v1.RelationshipFilter{
			ResourceType:       pr.ObjectType,
			OptionalResourceId: pr.Object,
		},
	}

	if pr.Relation != "" {
		req.RelationshipFilter.OptionalRelation = pr.Relation
	}

	if pr.SubjectType != "" {
		req.RelationshipFilter.OptionalSubjectFilter = &v1.SubjectFilter{
			SubjectType: pr.SubjectType,
		}
		if pr.Subject != "" {
			req.RelationshipFilter.OptionalSubjectFilter.OptionalSubjectId = pr.Subject
		}
		if pr.SubjectRelation != "" {
			req.RelationshipFilter.OptionalSubjectFilter.OptionalRelation = &v1.SubjectFilter_RelationFilter{
				Relation: pr.SubjectRelation,
			}
		}
	}

	if _, err := pa.permissionClient.DeleteRelationships(ctx, req); err != nil {
		return errors.Wrap(errRemovePolicies, handleSpicedbError(err))
	}
	return nil
}
```

rodneyosodo replied:
This automatically deletes the other group entities

arvindh123 commented on pkg/clients/status.go:
```diff
@@ -19,6 +19,8 @@ const (
 	// DisabledStatus represents disabled Client.
 	DisabledStatus
+	// DeletedStatus represents a client that will be deleted.
+	DeletedStatus
```
`DeleteStatus` can be highest status , If we want to hide the Deleted users from List with status=all `http://localhost/users?status=all`, 
If DeleteStatus becmes higher status, then superadmin can get deleted user by explicitly like `status=deleted`

rodneyosodo replied:
Highest status meaning 0 or n?

arvindh123 replied:
`n`

rodneyosodo committed:
feat(auth): delete policies in one req

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix(users): enable to change your status

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin commented on auth/domains.go:
```diff
@@ -199,4 +199,7 @@ type DomainsRepository interface {
 	CheckPolicy(ctx context.Context, pc Policy) error
+
+	// DeleteUserPolicies deletes user policies from domains database.
+	DeleteUserPolicies(ctx context.Context, id string) (err error)
```
Ranme to `DeletePolicies`.

rodneyosodo replied:
This is different from `DeletePolicies` will delete based on `subject_type`, `subject_id`, `subject_relation`, `object_type` and `object_id` while `DeleteUserPolicies` will delete based on `subject_id` which is `user_id`

dborovcanin commented on users/delete_handler.go:
```diff
@@ -0,0 +1,102 @@
+		return
+	}
+
+	if dbUsers.Total > defLimit {
```
Let's fetch and remove in batches, rather than fetching all first then removing, just to avoid edge cases where we slice of removed users are huge.

dborovcanin commented on users/delete_handler.go:
```diff
@@ -0,0 +1,102 @@
+	}()
+}
+
+func (h *handler) routine(ctx context.Context) {
```
Rename to `handle`.

dborovcanin commented on users/delete_handler.go:
```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
```
Add a description of this cron job explaining what it does and why. Keep in mind that scaling Users svc will result in scaling this cron, which is not ideal, and we will probably extract this to a separate process in the future.

rodneyosodo committed:
docs(delete-user): Add docs for DeleteHandler cron job

Add coumentation for the delete handler cron job based on what it does and why

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
